core/dataset.cpp

```diff
@@ -324,6 +324,37 @@ Dataset Dataset::slice(const Slice slice1, const Slice slice2,
   return Dataset{DatasetConstProxy(*this).slice(slice1, slice2, slice3)};
 }
 
+/// Rename dimension `from` to `to`.
+void Dataset::rename(const Dim from, const Dim to) {
+  if (m_dims.count(to) != 0)
+    throw except::DimensionError("Duplicate dimension.");
+  if (m_dims.count(from) != 1)
+    return;
+
+  const auto relabel = [from, to](auto &map) {
+    auto node = map.extract(from);
+    node.key() = to;
+    map.insert(std::move(node));
+  };
+  relabel(m_dims);
+  relabel(m_coords);
+  for (auto &item : m_coords)
```

Comment:
https://github.com/scipp/scipp/issues/248 is still open (operations with `Dataset` do not give good guarantees yet.

In this case however, `Variable::rename` should never fail, since all dimensions of variables are also part of `m_dims`, this this should fail before?